File: producer-consumer/src/main/java/com/wardziniak/kafka/config/ProducerConfigBuilder.java
package com.wardziniak.kafka.config;

import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.Optional;
import java.util.Properties;

/**
 * Created by wardziniak on 09.06.18.
 */
public class ProducerConfigBuilder implements ConfigBuilder {

    private static final String BOOTSTRAP_SERVERS_DEFAULT = "localhost:9092";
    private static final String KEY_SERIALIZER_DEFAULT = StringSerializer.class.getCanonicalName();
    private static final String VALUE_SERIALIZER_DEFAULT = StringSerializer.class.getCanonicalName();

    private Optional<String> bootstrapServer = Optional.empty();
    private Optional<String> keySerializer = Optional.empty();
    private Optional<String> valueSerializer = Optional.empty();

    public ProducerConfigBuilder withServer(String bootstrapServer) {
        this.bootstrapServer = Optional.ofNullable(bootstrapServer);
        return this;
    }

    private ProducerConfigBuilder withKeySerializer(String keySerialize) {
        this.keySerializer = Optional.ofNullable(keySerialize);
        return this;
    }

    private ProducerConfigBuilder withValueSerializer(String valueSerializer) {
        this.valueSerializer = Optional.ofNullable(valueSerializer);
        return this;
    }

    public ProducerConfigBuilder withSerializers(String keySerialize, String valueSerializer) {
        return this.withKeySerializer(keySerialize).withValueSerializer(valueSerializer);
    }

    @Override
    public Properties buildConfig() {
        Properties properties = new Properties();
        properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer.orElse(BOOTSTRAP_SERVERS_DEFAULT));
        properties.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, keySerializer.orElse(KEY_SERIALIZER_DEFAULT));
        properties.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, valueSerializer.orElse(VALUE_SERIALIZER_DEFAULT));
        return properties;
    }
}

File: producer-consumer/src/main/scala/com/wardziniak/kafka/scala/config/ProducerConfigBuilder.scala
package com.wardziniak.kafka.scala.config

import java.util.Properties

import org.apache.kafka.clients.producer.internals.DefaultPartitioner
import org.apache.kafka.clients.producer.{Partitioner, ProducerConfig}
import org.apache.kafka.common.serialization.StringSerializer

case class ProducerConfigBuilder(
  bootstrapServer: String = "localhost:9092",
  keySerializer: String = classOf[StringSerializer].getCanonicalName,
  valueSerializer: String = classOf[StringSerializer].getCanonicalName,
  partitioner: Partitioner = new DefaultPartitioner()
) extends ConfigBuilder {


  override def buildConfig = {
    val properties = new Properties
    properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer)
    properties.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, keySerializer)
    properties.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, valueSerializer)
    properties.setProperty(ProducerConfig.PARTITIONER_CLASS_CONFIG, partitioner.getClass.getCanonicalName)
    properties
  }
}
